hazelcast-jet-python/src/main/java/com/hazelcast/jet/python/PythonService.java

```diff
@@ -0,0 +1,199 @@
+import com.hazelcast.logging.ILogger;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.grpc.ManagedChannel;
+import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * The service object used by the "map using Python" pipeline stage. As a
+ * user you don't have to deal with this class directly. It is used when
+ * you write {@link #mapUsingPython
+ * stage.apply(PythonService.mapUsingPython(pyConfig))}
+ *
+ * @since 4.0
+ */
+public final class PythonService {
+
+    private static final String JET_TO_PYTHON_PREFIX = "jet_to_python_";
+    static final String MAIN_SHELL_SCRIPT = JET_TO_PYTHON_PREFIX + "main.sh";
+
+    private final ILogger logger;
+    private final JetToPythonServer server;
+    private final ManagedChannel chan;
+    private final StreamObserver<InputMessage> sink;
+    private final Queue<CompletableFuture<List<String>>> futureQueue = new ConcurrentLinkedQueue<>();
+
+    private final CountDownLatch completionLatch = new CountDownLatch(1);
+    private volatile Throwable exceptionInOutputObserver;
+
+    PythonService(Context procCtx, PythonServiceContext serviceContext) {
+        logger = procCtx.logger();
+        try {
+            server = new JetToPythonServer(serviceContext.runtimeBaseDir(), logger);
+            int serverPort = server.start();
+            chan = NettyChannelBuilder.forAddress("127.0.0.1", serverPort)
+                                      .usePlaintext()
+                                      .build();
+            JetToPythonStub client = JetToPythonGrpc.newStub(chan);
+            sink = client.streamingCall(new OutputMessageObserver());
+        } catch (Exception e) {
+            throw new JetException("PythonService initialization failed", e);
+        }
+    }
+
+    /**
+     * A stage-transforming method that adds a "map using Python" pipeline stage.
+     * Use it with {@code stage.apply(PythonService.mapUsingPython(pyConfig))}.
+     */
+    @Nonnull
+    public static FunctionEx<StreamStage<String>, StreamStage<String>> mapUsingPython(
+            @Nonnull PythonServiceConfig cfg
+    ) {
+        return s -> s
+                .mapUsingServiceAsyncBatched(factory(cfg), Integer.MAX_VALUE, PythonService::sendRequest)
+                .setName("mapUsingPython");
+    }
+
+    /**
+     * A stage-transforming method that adds a "map using Python" pipeline stage.
+     * Use it with {@code stage.apply(PythonService.mapUsingPythonBatch(pyConfig))}.
+     */
+    @Nonnull
+    public static FunctionEx<BatchStage<String>, BatchStage<String>> mapUsingPythonBatch(
```

Comment:
Isn't this method implementation is 1:1 copy of the one above ?

Reply:
It's not possible to make them have same name, as the type params are different

Reply:
The batch in the method name confused me, I've missed the signature